### PR TITLE
Combine global and suite configs

### LINKED ISSUE #3419: [New Config 1] Combine Global and Suite Config Spec
PR #3348 Sets out how an end point for unifying global and local configs, but contains many very complex changes.

An alternative first step to making this change in a way that produces working code is to combine global and local changes whilst making the bare minimum of changes to the code-base.

This issue should cover:
- [x] Move `cylc/flow/cfgspec/suite.py` -> `cylc/floc/config_schema.py`
- [x] Add items from `cylc/flow/cfgspec/global.py` to the new `config_schema.py`
- [x] Move the global config lazy loader (`cylc/flow/cfgspec/glbl_cfg.py`) to `cylc/flow/glbl_cfg_lazy.py`
- [x] Grep the rest of the repo for `cylc.flow.cfgspec` and ensure that the new spec is being imported in all cases.
- [x] Delete the `cfgspec` folder.
- [ ] :tada: Celebrate a small victory on the road to the replacing Rose Suite Run with a new Cylc command.

### PR DESCRIPTION
**Pull request by wxtim**
### Overview
Fix #3419 

PR #3348 Sets out how an end point for unifying global and local configs, but contains many very complex changes.

An alternative first step to making this change in a way that produces working code is to combine global and local changes whilst making the bare minimum of changes to the code-base.

### Todo list
This issue should cover:
- [x] Move `cylc/flow/cfgspec/suite.py` -> `cylc/floc/config_schema.py`
- [x] Add items from `cylc/flow/cfgspec/global.py` to the new `config_schema.py`
- [x] Move the global config lazy loader (`cylc/flow/cfgspec/glbl_cfg.py`) to `cylc/flow/glbl_cfg_lazy.py`
- [x] Grep the rest of the repo for `cylc.flow.cfgspec` and ensure that the new spec is being imported in all cases.
- [x] Delete the `cfgspec` folder.
- [ ] :tada: Celebrate a small victory on the road to the replacing Rose Suite Run with a new Cylc command.

### Notes on review
Reviewers should take a particular interest in corner cases where existing tests do not cover the workings of the combined configs.

@oliver-sanders & @hjoliver Tagged for normal review and sanity checking.
@dpmatthews Tagged because doing things in this order was his idea and to keep him in the loop.

### CODE REVIEW

File: cylc/flow/config_schema.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
```
Comment (wxtim) [inline cylc/flow/config_schema.py]: This is  the main changed file - it may be profitable for the reviewer to compare this:
```bash
git difftool upstream/master:cylc/flow/cfgspec/suite.py cylc/flow/config_schema.py
```

File: cylc/flow/config_schema.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+class GlobalConfig(ParsecConfig):
+    """
+    Handle global (all suites) site and user configuration for cylc.
+    User file values override site file values.
+    """
+
+    _DEFAULT = None
+    _HOME = os.getenv("HOME") or get_user_home()
+    CONF_BASENAME = "flow.rc"
+    SITE_CONF_DIR = os.path.join(os.sep, "etc", "cylc", "flow", CYLC_VERSION)
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/flow/config_schema.py]: Can we create an issue linked to the project?

File: cylc/flow/config_schema.py [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+    def __init__(self, fpath, output_fname, tvars):
+        """Return the default instance."""
+        ParsecConfig.__init__(
+            self, SPEC, upg, output_fname, tvars, cylc_config_validate
+        )
+        self.loadcfg(fpath, "suite definition")
+
+
+class GlobalConfig(ParsecConfig):
+    """
```
Comment (wxtim) [inline cylc/flow/config_schema.py]: Strategy - @oliver-sanders  is it worth doing the work to combine this with plain old config.py next? This feels like it might be a sensible strategy....

**Comment by kinow** [comment]
In Travis:

```
./tests/authentication/05-full-control.t                        (Wstat: 0 Tests: 12 Failed: 7)
  Failed tests:  6-12
```

But it is passing in my computer. Kicking Travis (gist with the error logs for the tests FWIW: https://gist.github.com/kinow/59dcc5d2f2134a0794cd039662195453)

File: cylc/flow/glbl_cfg_lazy.py [hunk 4]
```diff
@@ -18,5 +18,5 @@
 
 def glbl_cfg(cached=True):
     """Load and return the global configuration singleton instance."""
-    from cylc.flow.cfgspec.globalcfg import GlobalConfig
+    from cylc.flow.config_schema import GlobalConfig
     return GlobalConfig.get_inst(cached=cached)
```
Comment (kinow) [inline cylc/flow/glbl_cfg_lazy.py]: Shouldn't this file be kept in the `cfgspec` directory/module? Or is that module going to be removed later?

File: cylc/flow/config_schema.py (hunk 3 above)
Comment (oliver-sanders) [reply cylc/flow/config_schema.py]: If by `config.py` you mean `RawSuiteConfig` then maybe yes. We don't need two different classes to represent the same config, the only difference now is the loading logic.

File: cylc/flow/glbl_cfg_lazy.py (hunk 4 above)
Comment (oliver-sanders) [reply cylc/flow/glbl_cfg_lazy.py]: Has `cfgspec` not been removed yet?

Comment (wxtim) [reply cylc/flow/glbl_cfg_lazy.py]: It has been.

**Comment by wxtim** [comment]
The last two tests under flaky tests will never work under the new system. Unfortunately I need to do some more work on this.

File: cylc/flow/config_schema.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/config_schema.py]: https://github.com/cylc/cylc-flow/issues/3167

File: cylc/flow/config_schema.py [hunk 5]
```diff
@@ -508,16 +508,24 @@ def config_getter(output_fname, tvars, suite_fpath=None, user=True, site=True):
...
-    if site:
-        CONFIG_FILEPATHS.append(
-            (SITE_CONF_DIR / CONF_BASENAME, upgrader.SITE_CONFIG)
-        )
-    if user:
-        CONFIG_FILEPATHS.append(
-            (USER_CONF_DIR / CONF_BASENAME, upgrader.USER_CONFIG)
-        )
+    conf_path_str = os.getenv("CYLC_CONF_PATH")
+    if conf_path_str:
```
Comment (oliver-sanders) [inline cylc/flow/config_schema.py]: I think this section should be merged with `if user`.

File: cylc/flow/config_schema.py [hunk 6]
```diff
@@ -499,7 +505,7 @@ def get_config(output_fname, tvars, suite_fpath=None, user=True, site=True):
         A CylcConfig object.
     """
     _HOME = pathlib.Path.home() or get_user_home()
-    SITE_CONF_DIR = pathlib.Path(_HOME, "mock_cylc_global")
+    SITE_CONF_DIR = pathlib.Path(os.sep, 'etc', 'cylc', 'flow', CYLC_VERSION)
```
Comment (oliver-sanders) [inline cylc/flow/config_schema.py]: Don't need to use `os.sep`, pathlib can handle `/etc`.

(infact pathlib can handle `"/etc/cylc/flow/", CYLC_VERSION`)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/glbl_cfg_lazy.py  1
- cylc/flow/config_schema.py  14
         

Complexity decreasing per file
==============================
+ cylc/flow/config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4376089&bid=14851048)

**Comment by dpmatthews** [comment]
No longer sure we should do this
